### DISCUSSION TITLE
Color highlighting internally for custom component

### DISCUSSION THREAD

**Discussion by AckslD** [nvim-lualine/lualine.nvim#586] [Q&A] [answered]
I'm trying to understand how to highlight pieces of a custom component but not fully sure how to handle this properly. I was looking at how this is done for the `diagnostics` and just to try it out I tried something like (I extracted the relevant code):

```lua
local status_symbols = {
  unclean = '✘',
  clean = '✔',
}
local colors = {
  clean = '%#lualine_b_diagnostics_info_normal#',
  unclean = '%#lualine_b_diagnostics_error_normal#',
}
summary = {
  clean = 38
  unclean = 16,
}
local format_summary = function(summary)
  local pieces = {}
  for _, status in ipairs({'clean', 'unclean'}) do
    local num = summary[status]
    if num > 0 then
      table.insert(pieces, string.format('%s%s%s', colors[status], status_symbols[status], num))
    end
  end
  return table.concat(pieces, ' ')
end
```

where I put the `format_summary` in the `x` section.

![gscreenshot_2022-02-25-091146](https://user-images.githubusercontent.com/23341710/155679223-a8ef396b-ba13-4c8d-971b-771f967eadf0.png)

It does sort of work although there is a weird edge. And also I guess this is not the intended way since I'm pretty much just taking the highlighting from `diagnostics`.

Any advice how to do this properly would be highly appreciated :)

Thanks in advance!

Edit:

I also did try to use `create_component_highlight_group` and `component_format_highlight` similarly to `diagnostics` but couldn't fully figure out how this should be done.

**Comment by AckslD** [comment]
Just for context, this is the WIP plugin I'm working on which I trying to do this :)

https://github.com/AckslD/nvim-gfold.lua

**Comment by shadmansaleh** [comment, marked as answer]
You will have to use `create_component_highlight_group` and `component_format_highlight` pair. Then lualine will set correct background color based on section and mode .

You can follow the structure on diagnostics/diff components . For your case you'll need something like this
```lua
local M = require('lualine.component'):extend()
local highlight = require 'lualine.highlight'

function M:init(options)
  M.super.init(self, options)
  self.status_symbols = {
    unclean = '✘',
    clean = '✔',
  }

  self.colors = {
    clean = highlight.create_component_highlight_group(
      {fg =  '#48b0bd'}, 'summery_clean', self.options),
    unclean = highlight.create_component_highlight_group(
    {fg = '#e55561'}, 'summery_unclean', self.options)
  }
end

function M:update_status()
  local summary = { -- fetch actual summery
    clean = 38,
    unclean = 16,
  }
  local pieces = {}
  for _, status in ipairs({'clean', 'unclean'}) do
    local num = summary[status]
    if num > 0 then
      table.insert(pieces, string.format('%s%s%s',
        highlight.component_format_highlight(self.colors[status]),
        self.status_symbols[status], num))
    end
  end
  return table.concat(pieces, ' ')
end

return M
```

**Reply by AckslD** [reply]
Thanks a lot @shadmansaleh! And should then `M` be placed in the `lualine` section?

Edit: I mean something like:
```lua
require('lualine').setup({
  ...
  sections = {
    ...
    lualine_x = {
      ...
      require(<module>),
      ...
    },
    ...
  },
  ...
})
```
where `<module>` points to the above

**Reply by shadmansaleh** [reply]
That would work . But better if you place it in `lua/lualine/components/nvim-gfold.lua` in nvim-gfold repo.

Then users can just do
```lua
require('lualine').setup({
  ...
  sections = {
    ...
    lualine_x = {
      ...
      'nvim-gfold',
      ...
    },
    ...
  },
  ...
})
```

**Reply by AckslD** [reply]
Ah awesome :tada:

**Reply by shadmansaleh** [reply]
gfold looks quite interesting good luck :)

By the way does it conflict with project.nvim ?

**Reply by AckslD** [reply]
Thanks :) Not fully finished yet but I can let you know if you want to try it when I fixed the last things for a first release.

I don't think it conflicts with `project.nvim` in the sense that you can easily use both.

But if you're asking if they're doing different things, I think yes. Correct me if I'm wrong but `project.nvim` does not tell you the git status of a repo. My main motivation for `nvim-gfold` is to have a to easily see status of different repos and see which need attention (unpushed, unclean etc.

The plugin is actually just a simple interface to the `gfold` project which I recently found. I've been looking for a while for a tool which can give status information about my current repos and find `gfold` really nice cause it does one thing well :) I tried others like, `myrepos`, `gita` etc but was never fully happy and they try to do much more than I need.

**Reply by shadmansaleh** [reply]
Thanks. I'll know if you make a release post on reddit :)

> I don't think it conflicts with project.nvim in the sense that you can easily use both

That's what I was looking for .

**Reply by AckslD** [reply]
The above worked great btw, thanks again!

![gscreenshot_2022-02-26-140848](https://user-images.githubusercontent.com/23341710/155844747-11f111cc-ca0c-4ccf-965d-bcf4ade6891b.png)

**Comment by wookayin** [comment]
Also: if you don't want to create a new highlight group but instead use existing highlight groups, a lua function (or expression) can simply return a string with `%#HLName#` (:help 'statusline')

```lua
function custom_summary_component()
  local summary = { -- fetch actual summary
    clean = 38,
    unclean = 16,
  }
  local pieces = {}
  for _, status in ipairs({'clean', 'unclean'}) do
    local num = summary[status]
    if num > 0 then
      table.insert(pieces, string.format('%s%s%s',
        ('%#MyHighlightGroup_%s#'):format(status), -- <=================== see this line
        self.status_symbols[status], num))
    end
  end
  return table.concat(pieces, ' ')
end
```
